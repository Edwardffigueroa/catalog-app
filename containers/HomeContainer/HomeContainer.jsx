"use client";
import ProductCard from "@/components/ProductCard";
import SearchBar from "@/components/SearchBar";
import React, { useState, useEffect } from "react";
import styles from "../../app/page.module.scss";
import useDebouncer from "@/hooks/useDebouncer";
import { orderAlphabeticalTitle, orderByPrice } from "@/helpers/functions";
import { ORDER_BY, ORDER_DESC } from "@/utils/constants";
import PropTypes from "prop-types";
import { ProductTypes } from "@/types";

const HomeContainer = ({ products: serverProducts }) => {
  const [products, setProducts] = useState(serverProducts || []);
  const [allowInteraction, setAllowInteraction] = useState(false);
  const [filters, setFilters] = useState({
    searchProduct: "",
    orderBy: ORDER_BY.PRICE,
  });

  const debouncedSearchTerm = useDebouncer(filters.searchProduct, 500);

  useEffect(() => {
    setAllowInteraction(true);
  }, []);

  useEffect(() => {
    const results = serverProducts.filter((product) => {
      return product?.title
        ?.toLowerCase()
        .includes(debouncedSearchTerm?.toLowerCase());
    });
    setProducts(results);
  }, [serverProducts, debouncedSearchTerm]);

  useEffect(() => {
    if (filters.orderBy === ORDER_BY.A_TO_Z) {
      setProducts((prev) => [...orderAlphabeticalTitle(prev)]);
    }
    if (filters.orderBy === ORDER_BY.Z_TO_A) {
      setProducts((prev) => [...orderAlphabeticalTitle(prev, ORDER_DESC)]);
    }
    if (filters.orderBy === ORDER_BY.PRICE) {
      setProducts((prev) => [...orderByPrice(prev)]);
    }
  }, [serverProducts, filters.orderBy, debouncedSearchTerm]);

  const handlerOnChangeFilters = (e) => {
    setFilters((prev) => ({
      ...prev,
      [e.target.name]: e.target.value,
    }));
  };

  return (
    <>
      <h2 className={styles.h2} data-testid="heading">
        All products
      </h2>
      <SearchBar
        disabled={!allowInteraction}
        filters={filters}
        onChange={handlerOnChangeFilters}
      />
      <section className={styles.products_container}>
        {products.map((product) => (
          <ProductCard key={product.id} {...product} />
        ))}
      </section>
    </>
  );
};

HomeContainer.propTypes = {
  products: PropTypes.arrayOf(ProductTypes).isRequired,
};

export default HomeContainer;
