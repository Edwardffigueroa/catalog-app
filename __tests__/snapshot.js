import { render } from "@testing-library/react";
import HomeContainer from "@/containers/HomeContainer/HomeContainer";

it("renders homepage unchanged", () => {
  const { container } = render(<HomeContainer products={[]} />);
  expect(container).toMatchSnapshot();
});
