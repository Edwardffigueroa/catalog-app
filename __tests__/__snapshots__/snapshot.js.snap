// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders homepage unchanged 1`] = `
<div>
  <h2
    class="h2"
    data-testid="heading"
  >
    All products
  </h2>
  <div>
    <input
      class="input"
      data-testid="search-bar"
      name="searchProduct"
      placeholder="Type your product..."
      type="text"
      value=""
    />
    <div
      class="filters_container"
    >
      <label
        class="select_label"
        for="orderBy"
      >
        Order by:
      </label>
      <select
        class="select"
        name="orderBy"
      >
        <option
          value="price"
        >
          Lower Price
        </option>
        <option
          value="az"
        >
          A-Z
        </option>
        <option
          value="za"
        >
          Z-A
        </option>
      </select>
    </div>
  </div>
  <section
    class="products_container"
  />
</div>
`;
